File: agnt-plugin/synthetic.js
// PulseMesh JS — Deterministic Synthetic Fallback
// Port of Python synthetic.js — seeded PRNG for reproducible fallback data

import { TelemetrySeries } from './models.js';
import { mulberry32, gaussian } from './util.js';

export function syntheticSeries(profile, reason, maxPoints = 512) {
  const seed = hashString(`${profile.id}|${profile.provider}|${profile.variable || ''}`) >>> 0;
  const rng = mulberry32(seed);
  const n = Math.max(24, parseInt(maxPoints) || 256);
  const provider = profile.provider.toLowerCase();
  const variable = (profile.variable || profile.params.variable || provider).toLowerCase();

  let base, amp, pulse, unit;
  if (provider.includes('quake') || variable.includes('quake')) {
    base = 2.0; amp = 0.8; pulse = 2.2; unit = 'magnitude';
  } else if (provider.includes('air') || variable.includes('aqi') || variable.includes('pm2') || variable.includes('pm10')) {
    base = 35.0; amp = 10.0; pulse = 18.0; unit = 'index';
  } else if (provider.includes('goes') || provider.includes('solar') || variable.includes('xray')) {
    base = -6.2; amp = 0.4; pulse = 1.1; unit = 'log10 W/m^2';
  } else if (variable.includes('wind')) {
    base = 8.0; amp = 2.5; pulse = 4.0; unit = 'km/h';
  } else if (variable.includes('precip')) {
    base = 0.1; amp = 0.25; pulse = 1.0; unit = 'mm';
  } else if (variable.includes('pressure')) {
    base = 1012.0; amp = 4.0; pulse = 8.0; unit = 'hPa';
  } else {
    base = 20.0; amp = 5.0; pulse = 3.0; unit = 'synthetic';
  }

  const values = [];
  const times = [];
  for (let i = 0; i < n; i++) {
    const t = i / Math.max(1, n - 1);
    const wave = amp * Math.sin(2 * Math.PI * t);
    const transient = pulse * Math.exp(-0.5 * ((t - 0.62) / 0.06) ** 2);
    const noise = gaussian(rng) * Math.max(Math.abs(amp) * 0.08, 0.05);
    values.push(base + wave + transient + noise);
    times.push(`synthetic:${String(i).padStart(4, '0')}`);
  }

  return new TelemetrySeries({
    profileId: profile.id,
    provider: profile.provider,
    label: profile.label || `${profile.id} (synthetic)`,
    sensorName: `Synthetic fallback for ${profile.id}: ${reason}`,
    times,
    values,
    unit,
    usedLiveData: false,
    fallbackReason: reason,
  });
}

function hashString(s) {
  let h = 0;
  for (let i = 0; i < s.length; i++) {
    h = ((h << 5) - h + s.charCodeAt(i)) | 0;
  }
  return h;
}

File: src/pulsemesh/synthetic.py
from __future__ import annotations

import math
import random

from .models import TelemetryProfile, TelemetrySeries


def synthetic_series(profile: TelemetryProfile, reason: str, max_points: int) -> TelemetrySeries:
    seed = hash((profile.id, profile.provider, profile.variable)) & 0xFFFFFFFF
    rng = random.Random(seed)
    n = max(24, int(max_points))
    provider = profile.provider.lower()
    variable = (profile.variable or profile.params.get("variable") or provider).lower()

    if "quake" in provider or "quake" in variable:
        base, amp, pulse = 2.0, 0.8, 2.2
        unit = "magnitude"
    elif "air" in provider or "aqi" in variable or "pm2" in variable or "pm10" in variable:
        base, amp, pulse = 35.0, 10.0, 18.0
        unit = "index"
    elif "goes" in provider or "solar" in provider or "xray" in variable:
        base, amp, pulse = -6.2, 0.4, 1.1
        unit = "log10 W/m^2"
    elif "wind" in variable:
        base, amp, pulse = 8.0, 2.5, 4.0
        unit = "km/h"
    elif "precip" in variable:
        base, amp, pulse = 0.1, 0.25, 1.0
        unit = "mm"
    elif "pressure" in variable:
        base, amp, pulse = 1012.0, 4.0, 8.0
        unit = "hPa"
    else:
        base, amp, pulse = 20.0, 5.0, 3.0
        unit = "synthetic"

    values: list[float] = []
    times: list[str] = []
    for i in range(n):
        t = i / max(1, n - 1)
        wave = amp * math.sin(2.0 * math.pi * t)
        transient = pulse * math.exp(-0.5 * ((t - 0.62) / 0.06) ** 2)
        noise = rng.gauss(0.0, max(abs(amp) * 0.08, 0.05))
        values.append(base + wave + transient + noise)
        times.append(f"synthetic:{i:04d}")

    return TelemetrySeries(
        profile_id=profile.id,
        provider=profile.provider,
        label=profile.label or profile.id,
        sensor_name=f"Synthetic fallback for {profile.provider}",
        times=times,
        values=values[-max_points:],
        unit=unit,
        used_live_data=False,
        fallback_reason=reason,
        metadata={"synthetic_seed": seed},
    )
